benchmarks/microbenchmarks/utils.py

```diff
@@ -0,0 +1,380 @@
+        """Convert string precision to torch dtype"""
+        return getattr(torch, precision_str.split(".")[-1])
+
+    def to_dict(self) -> Dict[str, Any]:
+        """Convert config to dictionary for main function"""
+        return {
+            "quantization": self.quantization,
+            "m": self.m,
+            "k": self.k,
+            "n": self.n,
+            "high_precision_dtype": self.high_precision_dtype,
+            "compile": self.compile,
+            "compile_mode": "default",
+            "device": self.device,
+            "model_type": self.model_type,
+            "output_dir": self.output_dir,
+        }
+
+
+class ToyLinearModel(torch.nn.Module):
+    def __init__(self, k=64, n=32, dtype=torch.bfloat16):
+        super().__init__()
+        self.linear1 = torch.nn.Linear(k, n, bias=False).to(dtype)
+
+    def forward(self, x):
+        x = self.linear1(x)
+        return x
+
+
+class LNLinearSigmoid(torch.nn.Module):
+    def __init__(self, fc_dim1, fc_dim2, dtype=torch.bfloat16):
+        super().__init__()
+        self.ln = torch.nn.LayerNorm(fc_dim1, elementwise_affine=False)
+        self.fc = torch.nn.Linear(fc_dim1, fc_dim2, bias=False).to(dtype)
+        self.sigmoid = torch.nn.Sigmoid()
+
+    def forward(self, x):
+        x = self.ln(x)
+        x = self.fc(x)
+        x = self.sigmoid(x)
+        return x
+
+
+def get_default_device() -> str:
+    if torch.cuda.is_available():
+        return "cuda"
+    elif torch.xpu.is_available():
+        return "xpu"
+    elif torch.backends.mps.is_available():
+        return "mps"
+    else:
+        print("Warning: Running on CPU as no GPU support was found")
+        return "cpu"
+
+
+def ffn_only(mod, fqn):
+    return isinstance(mod, torch.nn.Linear) and "feed_forward" in fqn
+
+
+def not_ffn_only(mod, fqn):
+    return isinstance(mod, torch.nn.Linear) and not ffn_only(mod, fqn)
+
+
+def ffn_or_attn_only(mod, fqn):
+    return isinstance(mod, torch.nn.Linear) and (
+        "feed_forward" in fqn or "attention" in fqn
+    )
+
+
+def quantize_model(
+    model: torch.nn.Module,
+    quantization: str,
+    **kwargs,
+):
+    """Quantize a model inplace or return a new quantized model.
+
+    Args:
+        model (torch.nn.Module): model to be quantized
+        quantization (str): quantization method to be used
+        kwargs: additional arguments to be passed to the quantization method
+    """
+    if "int4wo" in quantization and not HAS_TRITON:
+        print("Warning: Triton not available, falling back to baseline")
+        return model
+
+    # Define kwargs
+    sparsity = kwargs.get("sparsity", None)
+    precision = kwargs.get("precision", None)
+
+    # Quantization techniques
+    if "baseline" in quantization:
+        return model
+    if "int8wo" in quantization:
+        quantize_(model, Int8WeightOnlyConfig())
+    if "int8dq" in quantization:
+        if sparsity and "semi" in sparsity:
```

Comment:
if this is doing "apply sparsity and then apply quantization", IMO it's cleaner to just have two config strings and call your API twice using the same function, instead of trying to branch here